### PR TITLE
Removed dependency on boost if compiler do support it

### LINKED ISSUE #122: Removing boost dependencies
Since boost can be a burdden (like on android), and since asio (http://think-async.com/Asio/) also exist without boost (being include-only library is even an added bonus). 
I've done a dirty work about this which seam to work as indented (not tested SSL yet) at my fork :
https://github.com/sebt3/Simple-Web-Server

I'm not issuing an PR since I'm guessing you would reject it :P

### PR DESCRIPTION
**Pull request by sebt3**
Fix #122 
As proposed in the issue.

### CODE REVIEW

**Comment by eidheim** [comment]
Great, I'll look at this next week (sorry, busy till then)!

File: client_http.hpp [hunk 1]
```diff
@@ -12,15 +11,40 @@
 #include <mutex>
 #include <type_traits>
 
+#ifdef USE_BOOST_ASIO
+#include <boost/asio.hpp>
+#define ASIO_NS boost::asio
```
Comment (314159) [inline client_http.hpp]: Why `#define ASIO_NS boost::asio`?  Why not use `using ASIO_NS = boost::asio;` instead?

Comment (sebt3) [inline client_http.hpp]: Pretty much to be style consistant with [REGEX_NS used on the server side](https://github.com/eidheim/Simple-Web-Server/blob/master/server_http.hpp#L38).

**Comment by sebt3** [comment]
@eidheim, expected, there's a lot to review in here ;)

**Comment by sebt3** [comment]
@eidheim up ?

**Comment by eidheim** [comment]
@sebt3 sorry for this taking so long, my plan is to finish review the PR Tomorrow or the day after. I also realised I had to read up on CMake since there were some changes there I did not understand 100% right away.

**Comment by sebt3** [comment]
Well FindAsio.cmake is the very standard way to find a library, and since it's a header only, I'm only looking for the header;
Many things from your original CMakeFile was moved to SimpleWebServer.cmake so it could be reused by anyone just by including it. All I did was to add support for c++1y (in a very standard way) which clang require to use experimental::filesystem. Now that everything reusable is in an includable file, the resulting CMakeFiles only contain instruction required to build the examples and tests.
Since you have a way to install your library, a FindSimpleWebServer.cmake (just like the FindAsio.cmake) might be a good idea. Yet I feel like most users of this project just copy the files they need in their sources, so the includable is better imho. You might want to provide both though ;)

**Comment by eidheim** [comment]
My initial thoughts are as follows:

* use for instance `namespace regex_ns = std;` instead of `#define REGEX_NS std`, somewhat like @314159 suggested.
  * I know libclang based IDE's such as juCi++ will be able to go to definition on `regex_ns` then, since macros can be problematic.
  * namespaces are typically lowercase

* According to http://think-async.com/Asio/AsioStandalone there is already a macro `ASIO_STANDALONE` for using standalone Asio. We should check for this macro definition in the header files. By default though, we should use `boost::asio` due to backward compability, but later we will probably use c++20 asio by default instead and that would most likely require a major version bump of Simple-Web-Server. 

* The example cpp files should stay as they are, thus keeping them easy to read, and we do not have to worry about c++17 filesystem yet. I think people will understand that the examples require boost, but the header files might be used without boost, especially those considering using standalone asio

* Keep the CMakeLists.txt files as they are. Keeping `cmake/Modules/FindAsio.cmake` would be nice though for people deciding to use standalone Asio, but the current CMakeLists.txt files are for building the examples that require boost asio. However, those that will use standalone Asio, would use their own CMakeLists.txt with the `ASIO_STANDALONE` macro set. 
  * Providing `cmake/Modules/SimpleWebServer.cmake` would be nice, but I think most people will use this library as a git submodule or simply copying the header files. Still, we could add cmake installation exports, but I would like to do this in a separate PR since it require a bit more work. I had a similar PR for this here: https://github.com/eidheim/tiny-process-library/pull/25, but decided to not add cmake installation exports due to added complexity, and because I am not completely convinced that this is something that is needed for smaller libraries (with only header files and/or one library file).

This said, I'm of course open for discussion regarding the above comments! And thank you again for this PR:) If you decide to keep your PR as is, I can do the above changes in a separate commit, but let us discuss it first of course.

**Comment by sebt3** [comment]
Replying dot per dot ;)
- As you please. As replied to @314159, I only choose the other way because you did already. I'll change this next week (no coding time planned this week)
- As per /usr/include/boost/asio/detail/config.hpp, there's also a `BOOST_ASIO_STANDALONE` that one may want to use to build using boost::asio, yet dont link with boost. So this point require some work to be done right. Beside, currently, I'm not setting this yet I'm still able to link without boost (since I'm only using the part not requiring boost). As I understand it, setting any of these 2 macros (depending if you include boost/asio or asio) only disable all the functions that takes boost typed arguments. Yet `ASIO_HEADER_ONLY` is the default behaviour...
- well std::experimental::filesystem is part of c++11 as far as I get it, it's going to be std::filesystem in c++17. c++1y is "preliminary" support for c++2017, just like c++0x was "preliminary" support for c++ 2011. The issue is while gcc implemented it in it's c++11 support, clang choose to not to (it wasnt mandatory to support those thing, being "experimental"), but provide an implementation in c++1y since they will provide an implementation in c++17 (as per spec). Anyway, what you suggest is actually removing features in an example :(
- That's actually the point of providing a `SimpleWebServer.cmake` instead of a `FindSimpleWebServer.cmake`. So the git submodule user can easily include the file and so if the MACRO requirement change keep being updated on that front too. My changes doesnt look to be in the same idea as the PR you linked. If you want me to bring this in a different PR, then, I'll want this PR 1st. I've only done within this PR, because my patch made it clear having no way to change the macro definitions on the users is a big problem : Anyone just updating the headers without changing his buildsystem to reflect the macro requirement will be at a surprise : non-boost is the default code wise, so my patch is not backward compatible as-is...

**Comment by Type1J** [comment]
I really like this move to standalone asio. Good job @sebt3!

**Comment by eidheim** [comment]
@sebt3 In hindsight I should have numbered the points. But regarding the two last points:

3. since boost::property_tree is used in the examples and thus boost is needed anyway, thus I do not see the need to add additional complexity here through additional ifdefs. When we move to c++17, on the other hand, we can also move to std::filesystem (skipping std::experimental::filesystem altogether). But this is some years into the future. I do not see that I suggest removing features in the examples here?
4. I think maybe we can keep FindSimpleWebServer.cmake, your argument of making it easier for people using this library as a git submodule is quite valid. However, for the time being, boost::asio should be the default option, so that we do not need to bump the major version just yet. There are a couple more things I plan to do before that would happen. That is why I suggest that the asio library should be used in the header files if `ASIO_STANDALONE` is set, and the boost::asio library otherwise. Not sure if we need to take into account `BOOST_ASIO_STANDALONE` too? Not yet maybe, as we should focus on standalone asio and boost::asio first and foremost.

**Comment by sebt3** [comment]
Ok, I'll work on a new PR based on your ideas. I'll try to split my commit in a more readable way though.
This will need wait until next week at best.

**Pull request closed without merging**